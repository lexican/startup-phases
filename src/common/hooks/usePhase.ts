import { IPhase } from 'src/common/types/types';
import { fetcher } from 'src/common/utils/fetcher';
import useSWR from 'swr';

interface IData {
  phases: IPhase[];
}

interface IUsePhase {
  data: IData | undefined;
  error: any;
}

export const usePhase = (): IUsePhase => {
  const { data, error } = useSWR<IData>('/api/phase', fetcher);

  return { data, error };
};
