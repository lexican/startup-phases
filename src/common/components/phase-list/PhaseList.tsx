import { usePhase } from '../../hooks/usePhase';
import { LoaderStyles } from '../loader/LoaderStyles';
import Phase from '../phase/Phase';
import Task from '../task/Task';
import { PhaseListStyles } from './PhaseListStyles';

const PhaseList = (): JSX.Element => {
  const { data } = usePhase();
  if (!data) {
    return <LoaderStyles />;
  }

  return (
    <PhaseListStyles>
      {data.phases.map((phase, index) => {
        const isTaskDisabled =
          index === 0 ? false : !data.phases[index - 1].isCompleted;
        return (
          <div key={phase.id}>
            <Phase
              title={phase.title}
              phaseNumber={index + 1}
              isCompleted={phase.isCompleted}
            />
            <ul>
              {phase.tasks.map(task => {
                return (
                  <li key={task.id}>
                    <Task
                      task={task}
                      phaseId={phase.id}
                      disabled={isTaskDisabled}
                    />
                  </li>
                );
              })}
            </ul>
          </div>
        );
      })}
    </PhaseListStyles>
  );
};

export default PhaseList;
