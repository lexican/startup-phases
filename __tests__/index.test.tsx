import { render, screen } from '@testing-library/react';
import PhaseList from 'src/components/phase-list/PhaseList';
import { usePhase } from 'src/hooks/usePhase';
import Home from '../pages/index';

jest.mock('src/common/hooks/usePhase', () => ({
  usePhase: jest.fn()
}));

jest.mock('src/common/components/loader/Loader', () => ({
  __esModule: true,
  default: () => <div data-testid="loader" />
}));

jest.mock('src/common/components/phase/Phase', () => ({
  __esModule: true,
  default: ({ isCompleted }: any) => (
    <div data-testid="phase">{isCompleted.toString()}</div>
  )
}));

jest.mock('src/common/components/random-fact/RandomFact', () => ({
  __esModule: true,
  default: () => <div data-testid="random-fact" />
}));

jest.mock('src/common/components/task/Task', () => ({
  __esModule: true,
  default: ({ onChange, checked, disabled }: any) => (
    <input
      data-testid="checkbox"
      type="checkbox"
      onChange={onChange}
      checked={checked}
      disabled={disabled}
    />
  )
}));

const phasesData = [
  {
    id: 1,
    title: 'Phase 1',
    isCompleted: true,
    tasks: [
      {
        id: 1,
        title: 'First task',
        isChecked: false
      },
      {
        id: 2,
        title: 'Second task',
        isChecked: false
      },
      {
        id: 3,
        title: 'Third task2',
        isChecked: false
      }
    ]
  },
  {
    id: 2,
    title: 'Phase 2',
    isCompleted: false,
    tasks: [
      {
        id: 1,
        title: 'First task',
        isChecked: false
      }
    ]
  }
];

describe('Home', () => {
  beforeEach(() => {
    (usePhase as jest.Mock).mockReturnValue({
      data: { phases: phasesData },
      updateTask: jest.fn()
    });
  });
  it('should render without errors', () => {
    render(<Home />);
  });

  it('should render home page header correctly', () => {
    render(<Home />);
    const header = screen.getByText('My startup progress');
    expect(header).toBeInTheDocument();
  });

  it('should render home page footer text correctly', () => {
    render(<Home />);
    const header = screen.getByText('AUDITION ASSIGNMENT');
    expect(header).toBeInTheDocument();
  });

  it('should render loader when data is null', () => {
    (usePhase as jest.Mock).mockReturnValue({
      data: null,
      updateTask: jest.fn()
    });

    render(<PhaseList />);

    const loader = screen.getByTestId('loader');
    expect(loader).toBeInTheDocument();
  });

  it('should render correct number of phases', () => {
    render(<PhaseList />);

    const phaseTitles = screen.getAllByTestId('phase');
    expect(phaseTitles).toHaveLength(phasesData.length);
  });

  it('should render correct number of checkboxes', () => {
    render(<PhaseList />);

    const checkboxes = screen.getAllByTestId('checkbox');
    expect(checkboxes).toHaveLength(4);
  });
});
