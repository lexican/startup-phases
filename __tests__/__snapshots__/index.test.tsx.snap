// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchyanged 1`] = `
<div>
  <div
    class="sc-gKPRtg cxZQAE"
  >
    <div>
      failed to load
    </div>
    <div
      class="container"
    >
      <div
        class="inner-container"
      >
        <h1>
          My startup progress
        </h1>
        <div
          class="sc-gswNZR hMTrYV"
        />
      </div>
    </div>
    <div
      class="footer"
    >
      <h2>
        AUDITION ASSIGNMENT
      </h2>
      <div
        class="sc-bcXHqe Kgrez"
      >
        <div
          class="dot"
        />
        <div
          class="dot"
        />
        <div
          class="dot-white"
        />
      </div>
    </div>
  </div>
</div>
`;
